src/hotspot/cpu/riscv/assembler_riscv.cpp

```diff
@@ -463,186 +463,169 @@ void Assembler::li(Register d, void* addr) {
   li(d, (long)(unsigned long)addr);
 }
 
-void Assembler::li(Register d, long imm) { // TODO optimize
-  unsigned long uimm = imm;
 
-  // Accurate copying by 11 bits.
-  int remBit = ((uimm & 0xffffffff00000000) != 0) ? 53 : 21;
-  short part = (uimm >> remBit) & 0x7FF;
-  addi(d, R0_ZERO, part);
-  slli(d, d, 11);
+// load 64-bit immediate value
+void Assembler::li(Register d, long imm) {
+  // tty->print_cr("li %s, 0x%lx at %p", d->name(), (unsigned long)imm, pc());
 
-  for (remBit -= 11; remBit > 0; remBit -= 11) {
-    part = (uimm >> remBit) & 0x7FF;
-    if (part != 0) addi(d, d, part);
-    slli(d, d, remBit >= 11 ? 11 : remBit);
+  if (-0x800 <= imm && imm < 0x800) {
+    addi(d, R0_ZERO, imm);
+    return;
   }
 
-  part = uimm & ((1 << (remBit + 11)) - 1);
-  if (part != 0) addi(d, d, part);
-}
+  long off = imm - (long)pc();
 
-// Load a 64 bit constant, optimized, not identifyable.
-// Tmp can be used to increase ILP. Set return_simm16_rest=true to get a
-// 16 bit immediate offset.
-int Assembler::load_const_optimized(Register d, long x, Register tmp, bool return_simm16_rest) {
-  // Avoid accidentally trying to use R0 for indexed addressing.
-  assert_different_registers(d, tmp);
+  if (INT32_MIN <= off && off <= INT32_MAX) {
+    // load using AUIPC
+    unsigned long uoff = off;
+    unsigned long low = uoff & 0xfff;
+    unsigned long high = (uoff >> 12) & 0xfffff;
+    if (low >= 0x800) {
+      ++high;
+    }
+    auipc(d, high);
+    if (low) {
+      addi(d, d, low);
+    }
+    return;
+  }
 
-  short xa, xb, xc, xd; // Four 16-bit chunks of const.
-  long rem = x;         // Remaining part of const.
+  unsigned long uimm = imm;
+  unsigned long value = uimm;
 
-  xd = rem & 0xFFFF;    // Lowest 16-bit chunk.
-  rem = (rem >> 16) + ((unsigned short)xd >> 15); // Compensation for sign extend.
+  unsigned long sha = 0; // shift amount for final value
+  if (imm < INT32_MIN || imm > INT32_MAX) {
+    while (!(value & 1)) {
+      ++sha;
+      value >>= 1;
+    }
+  }
 
-  if (rem == 0) { // opt 1: simm16
-    li_PPC(d, xd);
-    return 0;
+  unsigned long low = value & 0xfff; // low section is 12 lowest bits
+  unsigned long mid = value >> 12; // mid section is 20+ following bits
+  if (low >= 0x800) {
```

Comment:
12 and 0xfff are repeated several times.
And 0xfff is (1 << 12) -1.

It makes sense to keep this knowledge in single place. 
For example:

const int lowest_const_bits = 12;
const int mask_lowest_const_bits = (1 << lowest_const_bits) - 1 ;

Reply:
May be there this such kind of optimized function already in OpenJDK codebase ?